Release/0.3.0

slama0077 opened the pull request:
## Summary
I've connected Dongha's PR with the changes I'd made with mine. This is a rough draft. There are optiomizations need to be made, specially in the read_file_v3 function. I also had to add a line in compute.py to drop a column in param_df, which I think is just a temporary fix for now. This PR is just for y'all to see what i've done and suggest any thing that needs to be changed. It is not ready to be merged at all. More refining is needed.

slama0077 committed:
Works with both version, version 3 and version 2.2

slama0077 committed:
Works for both version version3 and version2.2

kumdonoaa commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -201,23 +268,23 @@ def node_key_func(x):
     else:
         raise RuntimeError("Unsupported file type: {}".format(file_type))
-    
-    return flowpaths, lakes, network, nexus
```
I think naming "df" instead of flowpaths or flowlines would make the code less readable.  Any better name or use either name based on the version of gpkg?

kumdonoaa commented:
Changing the content of def read_geopkg drastically would cause some confusion (legacy code or backward compatibility). Can we keep the function intact while creating an alternative function such as def read_geopkg_dev?  We should keep the original df read_geopkg until we finalize the code development?

slama0077 commented:
On it, right now!

slama0077 commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -84,36 +105,82 @@ def read_layer(layer_name):
+            )
+            #merging on flowpath df to get mainstem attribute
+            df = pd.merge(df, flowpaths_df[['flowpath_id', 'mainstem']], on='flowpath_id', how='outer')
+        elif not flowlines_df.empty:
```
@kumdonoaa, arent we suppose to raise error if either one of flowlines_df or flowlines_attribute is missing? Is routing even possible if one of them is empty?

slama0077 committed:
Used Multiprocessing to read the channelrouting files to build the qlateral

Increased the speed by about 5 times for Forcing Array Construction

slama0077 committed:
Combined Dongha's new code with mine and made it run for version 3. Qlateral of the nexus is passed to downstream flowline.

crosswaslking nexus, poi is also done

if the qlateral_input files is channel routing, then it follows old convention of routing.

added multiprocessing in reading channel routing files which helps to create lateral array 5 times faster

headwater routing needs to be done with interpolation

slama0077 committed:
Cleared some test garbage

quinnylee commented on src/troute-config/troute/config/network_topology_parameters.py:
```diff
@@ -218,6 +220,14 @@ class Columns(BaseModel):
     """
     mainstem ID
     """
+    flowpath_id: Optional[str]
```
The addition of these lines causes an error when routing over an NHD hydrofabric:

```
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/NHDNetwork.py", line 169, in read_geo_file
    self._dataframe = self.dataframe[list(cols.values())]
  File "/Users/qylee/Documents/t-route-fork/.venv/lib/python3.10/site-packages/pandas/core/frame.py", line 4096, in __getitem__
    indexer = self.columns._get_indexer_strict(key, "columns")[1]
  File "/Users/qylee/Documents/t-route-fork/.venv/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6199, in _get_indexer_strict
    self._raise_if_missing(keyarr, indexer, axis_name)
  File "/Users/qylee/Documents/t-route-fork/.venv/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6251, in _raise_if_missing
    raise KeyError(f"{not_found} not in index")
KeyError: '[None] not in index'
```

What I think is happening here is that because the NHD network topology parameters don't contain `flowpath_id` and `flowpath_toid`, they automatically get set to `None`, which makes pandas unhappy. A potential fix to this is already hinted at in `troute-network/troute/NHDNetwork.py`, lines 158-160, where someone seemed to run into a similar issue when implementing support for the HYFeatures network.
```
# Remove 'mainstem' col if it exists:
if 'mainstem' in cols:
    del cols['mainstem']
```

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -405,15 +579,47 @@ def preprocess_network(self, flowpaths, nexus):
+            # Instead, self._nexus_to_reach is used to correctly link the lateral flow at a nexus node 
+            # to its downstream reach.            
+            # make the flowpath linkage, ignore the terminal nexus              
+            self._flowpath_dict = {}
```
The definition of `self._flowpath_dict` is now locked behind this `if` statement and will never get defined for HF v2.2, causing this issue in line 1033:
```
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/HYFeaturesNetwork.py", line 1034, in build_qlateral_array
    qlats_df.rename(index=self.downstream_flowpath_dict, inplace=True)
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/HYFeaturesNetwork.py", line 519, in downstream_flowpath_dict
    return self._flowpath_dict
AttributeError: 'HYFeaturesNetwork' object has no attribute '_flowpath_dict'
```

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -789,27 +1005,34 @@ def process_file(f):
+                end_time = time.time()
+                print(f"Parallel read time for {len(qlat_files)} files: {end_time - start_time} seconds")
+                
+            qlats_df = lateralflows_df             
```
Something strange has happened with the way `lateralflows_df` is generated. Here is what it looks like in the `ngiab` branch:
<img width="993" height="210" alt="Screenshot 2025-10-24 at 2 11 18 PM" src="https://github.com/user-attachments/assets/6029be98-3cb4-4a60-9b64-2471ce26b7a0" />

And here is what it looks like in this branch (the `feature_id` column gets copied a lot):
<img width="804" height="203" alt="Screenshot 2025-10-24 at 2 34 23 PM" src="https://github.com/user-attachments/assets/1045e623-43fc-4eea-9cd6-077c0322d6e8" />

I'm not sure if this will cause issues, but I felt like it was worth noting.

quinnylee replied:
Clarification: this happens with the HF v2.2

quinnylee reviewed (changes requested):
Once these changes are addressed, I'll test again for backwards compatibility

slama0077 committed:
Completed Head Water Interpoltion

Removed flowpath attributes manually if the network type is NHD

slama0077 committed:
Fix the broken lateralflow dataframe is version 2.2. Forgot to set feature_id as index in other file types

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -115,6 +118,161 @@ def read_layer(layer_name):
+        version_tag = 'v22'
+        layers_to_read = ['flowpaths', 'flowpath_attributes']
+    else:
+        raise runtimeError(
```
should be `RuntimeError` (capital R)

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -198,26 +356,25 @@ def node_key_func(x):
-        raise RuntimeError("Unsupported file type: {}".format(file_type))
-    
-    return flowpaths, lakes, network, nexus
+        raise runtimeError("Unsupported file type: {}".format(file_type))
```
should be `RuntimeError` (capital R)

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -968,20 +1241,62 @@ def read_preprocessed_data(self,):
+        df = df.reset_index()[['feature_id', 'q_lateral']]
         df.rename(columns={'q_lateral': f'{ts}'}, inplace=True)
         df.index.name = None
+        assert df["feature_id"].is_unique, f"'feature_id's must be unique. '{f!s}' contains duplicate 'feature_id's: {pformat(df.loc[df['feature_id'].duplicated(), 'feature_id'].to_list())}"
```
Is this f-string formatted correctly?

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -968,20 +1241,62 @@ def read_preprocessed_data(self,):
+    if extension=='.csv':
+        df = pd.read_csv(file_name)
+        df['feature_id'] = df['feature_id'].map(lambda x: int(str(x).removeprefix('nex-')) if str(x).startswith('nex') else int(x))
+        assert df["feature_id"].is_unique, f"'feature_id's must be unique. '{f!s}' contains duplicate 'feature_id's: {pformat(df.loc[df['feature_id'].duplicated(), 'feature_id'].to_list())}"
```
Is this f-string formatted correctly?

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -968,20 +1241,62 @@ def read_preprocessed_data(self,):
+        df = pq.read_table(file_name).to_pandas().reset_index()
+        df.index.name = None
+        df['feature_id'] = df['feature_id'].map(lambda x: int(str(x).removeprefix('nex-')) if str(x).startswith('nex') else int(x))
+        assert df["feature_id"].is_unique, f"'feature_id's must be unique. '{f!s}' contains duplicate 'feature_id's: {pformat(df.loc[df['feature_id'].duplicated(), 'feature_id'].to_list())}"
```
Is this f-string formatted correctly?

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -968,20 +1241,62 @@ def read_preprocessed_data(self,):
+        df.rename(columns={'q_lateral': f'{ts}'}, inplace=True)
+        df.index.name = None
+        df['feature_id'] = df['feature_id'].map(lambda x: int(str(x).removeprefix('nex-')) if str(x).startswith('nex') else int(x))
+        assert df["feature_id"].is_unique, f"'feature_id's must be unique. '{f!s}' contains duplicate 'feature_id's: {pformat(df.loc[df['feature_id'].duplicated(), 'feature_id'].to_list())}"
```
Is this f-string formatted correctly?

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -250,12 +407,55 @@ def load_bmi_data(value_dict, bmi_parameters,):
+    headwater_reaches = dataframe[~dataframe.index.isin(dataframe.downstream)].index.to_list()
+    network_mod = network_mod.drop_duplicates(subset=['nexus_id']).reset_index(drop=True).set_index("nexus_id", drop = True)
+
+    big_dict = {}
```
perhaps a more descriptive name for `big_dict` and `small_dict` would be useful

quinnylee reviewed (changes requested):
I checked backwards compatibility with NHD and HYFeatures 2.2 and checked parity in routed flows. All of that looks good! Routing also works for HYFeatures 3.0

Only thing is a couple of formatting issues

quinnylee commented:
> I checked backwards compatibility with NHD and HYFeatures 2.2 and checked parity in routed flows. All of that looks good! Routing also works for HYFeatures 3.0
> 
> Only thing is a couple of formatting issues

NHD runs are exactly the same in the original vs this branch. Here is the extent of the difference between HYFeatures 2.2 (no DA, Lower Colorado test)
this branch:
<img width="262" height="115" alt="Screenshot 2025-10-28 at 10 43 32 AM" src="https://github.com/user-attachments/assets/802b5743-08ed-4d05-afeb-8ecf21b55b95" />
original:
<img width="340" height="156" alt="Screenshot 2025-10-28 at 10 45 40 AM" src="https://github.com/user-attachments/assets/4a756935-b6a6-472a-922e-8bd277602554" />

(with DA, Lower Colorado test)
this branch:
<img width="264" height="83" alt="Screenshot 2025-10-28 at 10 46 52 AM" src="https://github.com/user-attachments/assets/682595d9-4e4b-41ca-8c05-9d0408186bdb" />
original:
<img width="339" height="117" alt="Screenshot 2025-10-28 at 10 46 24 AM" src="https://github.com/user-attachments/assets/b8ba36b4-e139-48e3-89b9-307a62d31fc5" />

Overall very similar, but with very small differences. I don't know if that's the extent of the difference you expect with changes to headwater routing

slama0077 committed:
empty commit

slama0077 committed:
changed compiler.sh

slama0077 committed:
skip the headwater interpolation for potential headreaches that flows into terminal nexus

Fixed f-string formatting

descriptive dictionary variable

slama0077 committed:
This commit should route v2.2 hydrofabric the new way

quinnylee commented on compiler.sh:
```diff
@@ -34,6 +34,7 @@ fi
 #export LIBRARY_PATH=<paths>:$LIBRARY_PATH
 #if you have custom dynamic library paths, uncomment below and export them
 #export LD_LIBRARY_PATHS=<paths>:$LD_LIBRARY_PATHS
+export NETCDF="/usr/local/include/"
```
Do we want to hardcode this path in the compiler? It seems to be different for all computers, like on my ubuntu machine it's `/usr/local/bin/`

There are instructions in the docs for how to find the appropriate path and how to change it though, so maybe it's not a big deal.

kumdonoaa replied:
I am not sure that I accidently git push my complier.sh, but hardcoded generally not good idea if it cause an issue to other's system. It may be commented out with general instruction when it can or should be used?

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -1080,10 +1080,10 @@ def process_file(f):
-                # Take flowpath ids entering NEXUS and replace NEXUS ids by the upstream flowpath ids
-                # version3.0 should be unaffected by this as _flowpath_dict is empty            
-                qlats_df.rename(index=self.downstream_flowpath_dict, inplace=True)
+            # if qlat_file_pattern_filter != "cat-*":
```
This changes routed outputs pretty significantly (I guess that is what you want.) 
Test case 1: the provided LowerColoradoTX_v4_noDA test, v2.2
- public t-route version
<img width="526" height="428" alt="Screenshot 2025-10-29 at 3 51 47 PM" src="https://github.com/user-attachments/assets/6618a4ee-8aa3-4dc7-8bb5-addf7ae67cbc" />

- this dev branch
<img width="282" height="299" alt="Screenshot 2025-10-29 at 3 41 28 PM" src="https://github.com/user-attachments/assets/b9dc06a0-f2c9-4add-aa28-eaf736e1a2c8" />

Test case 2: the provided LowerColoradoTX_v4 test, hf v2.2
- public t-route version:
<img width="355" height="114" alt="Screenshot 2025-10-29 at 3 55 03 PM" src="https://github.com/user-attachments/assets/8f00b515-70af-4b2a-89b1-27b3d8cd8a0d" />

- this dev branch:
<img width="266" height="91" alt="Screenshot 2025-10-29 at 3 54 56 PM" src="https://github.com/user-attachments/assets/83671c9e-2529-4481-b8ab-9608e1d3ccfd" />


validation of results with gaged data to come

quinnylee replied:
Validation attempt for LowerColorado TX test:
outputs are shown for catchment closest to gage-08117995 (colorado river at gail, tx)

dev branch results:
<img width="535" height="193" alt="Screenshot 2025-10-29 at 4 30 27 PM" src="https://github.com/user-attachments/assets/46076e8c-7164-4f5f-ba79-1c262917ea18" />


public t-route results: 
<img width="604" height="262" alt="Screenshot 2025-10-29 at 4 30 31 PM" src="https://github.com/user-attachments/assets/1a2974ca-37c7-4fe4-8f7b-b8f1b6ac3e2f" />


USGS discharge:
<img width="963" height="718" alt="Screenshot 2025-10-29 at 4 32 01 PM" src="https://github.com/user-attachments/assets/e92271e5-d743-4b10-9f24-30f36c202ed0" />

So I guess nothing really lined up, not really sure what this little test tells us

quinnylee commented:
Clarification on previous testing: All testing done on LowerColorado_v4 was actually on hf 2.1, not 2.2. Which is still useful, but now I need to start testing on hf 2.2.

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -405,15 +627,47 @@ def preprocess_network(self, flowpaths, nexus):
+            keep = (~id_num.isna()) & (~toid_num.isna()) & (~id_num.isin(self.terminal_nexus))
+            self._nexus_to_reach = dict(zip(id_num[keep].astype(int), toid_num[keep].astype(int))) 
+        else:
+            id_num   = pd.to_numeric(nexus["id"].astype(str).str.extract(r"(\d+)", expand=False), errors="coerce")
```
For HFs v2.1 and v2.2, if we turn off waterbodies, the nexus layer in the geopackage isn't read. So this line breaks the routing run since the `nexus` dataframe is empty. You should be able to read ids by manipulating the `wb-` string in the `flowpaths` layer.

slama0077 committed:
Make the code always read nexus layer

quinnylee commented on src/troute-network/troute/HYFeaturesNetwork.py:
```diff
@@ -632,10 +890,13 @@ def preprocess_waterbodies(self, lakes, nexus):
+        break_network_at_waterbodies = self.waterbody_parameters.get(
+            "break_network_at_waterbodies", False
+        )
+        if not network.empty and break_network_at_waterbodies:
```
This line breaks DA for HYFeatures-based runs if waterbodies are set to False. `self._gages` is empty if `break_network_at_waterbodies` is False, leading to this error
```
File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-nwm/src/nwm_routing/__main__.py", line 132, in main_v04
    data_assimilation = DataAssimilation(
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/DataAssimilation.py", line 990, in __init__
    NudgingDA.__init__(self, network, from_files, value_dict, da_run)
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/DataAssimilation.py", line 216, in __init__
    self._usgs_df = _create_usgs_df(data_assimilation_parameters, streamflow_da_parameters, run_parameters, network, da_run)
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/DataAssimilation.py", line 1137, in _create_usgs_df
    nhd_io.get_obs_from_timeslices(
  File "/Users/qylee/Documents/t-route-fork/t-route/src/troute-network/troute/nhd_io.py", line 1184, in get_obs_from_timeslices
    df[crosswalk_gage_field] = np.asarray(df[crosswalk_gage_field]).astype('<U15')
  File "/Users/qylee/Documents/t-route-fork/.venv/lib/python3.10/site-packages/pandas/core/frame.py", line 4090, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/Users/qylee/Documents/t-route-fork/.venv/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 3809, in get_loc
    raise KeyError(key) from err
KeyError: 'gages'
```

Tested on `LowerColorado_TX_v4/test_AnA_V4_HYFeature.yaml` but with `break_network_at_waterbodies` set to False

slama0077 committed:
Updated the code to create gage dataframes regardless of what break_network_bodies parameter is. But I think the better way to do this in Hyfeature implementation is to seperate the creation of lake_gage_df from the normal gage dataframe because creation of lake_gage_df should certainly depend on break_network_waterbody, and that's how NHD functions.

quinnylee reviewed (approved):
I'm sure I've missed some niche use case of t-route, but this passes all the tests I came up with. It might be worth considering splitting all of the v3.0 changes into a separate .py file one day, since there are quite a few if/else statements, but I think this works for now.
